Client/Frontend/Home/PanelDataObservers.swift

```diff
@@ -49,21 +49,31 @@ class ActivityStreamDataObserver: DataObserver {
     }
     
     func invalidate(highlights: Bool) {
+        guard !profile.isShutdown else {
+            return
+        }
+
         self.delegate?.willInvalidateDataSources()
 
         let notify = {
             self.delegate?.didInvalidateDataSources()
         }
-        
+
         let invalidateTopSites: () -> Success = {
             self.profile.history.setTopSitesNeedsInvalidation()
             return self.profile.history.updateTopSitesCacheIfInvalidated() >>> succeed
```

Comment:
Consider moving `uTSCIF` and friends from `BrowserHistory` into `HistoryRecommendations`, and then figuring out a combination of the two that runs a single transaction.